### ISSUE TITLE
BE: Events: Add "national outreach calendar" checkbox to Event UI

### ISSUE THREAD

**Issue by jilladams** [department-of-veterans-affairs/va.gov-cms#10089]
**STATUS**
- design done in #14788 
- need to form change management plan (if any) prior to implementation (ask DaveC at 9/25 product sync)

### User story
AS AN Outreach Events editor (and only outreach events editor) 
I WANT my event to get posted on the Outreach Events calendar by default (without any specific input from me)
SO THAT I don't have any new cognitive load added.

AS AN Editor who can create events for a Facility
I WANT to be given the option of publishing to the Outreach Events calendar (i.e. "national" calendar) in addition to my usual options
SO THAT I can easily expose an event with broad appeal and remote accessibility to all Veterans.

## Notes / background

**Current Event creation form:**
https://prod.cms.va.gov/node/add/event

**Design**
[Figma file](https://www.figma.com/file/8ZT1nnu8r2twbOlEhysnNp/Events?type=design&node-id=0%3A1&mode=design&t=EsFeeJGKPh7v29n8-1)

### Tasks
- Create checkbox field (non-required)
- BE processing to add Event to outreach hub cal – OE Cal is a specific node, an Event List: https://prod.cms.va.gov/outreach-and-events/events
  - "spike" chunk of work needed to figure out how Event Lists get turned into calendars on the FE
- Inspect FE template to ensure it can handle an event that's assigned to multiple Event Lists
- Implement hiding logic for select users

## Acceptance Criteria
- [x] Implementation of checkbox follows Figma design above
- [x] On Event edit UI, any Facility Editor can check a box to also publish to the Outreach events calendar
- [x] For an Editor whose only section is the Outreach Hub, the checkbox is hidden
- [x] On Save with the checkbox selected, the Event is added to the selected facility _and_ to the Outreach calendar
- [x] **DO NOT MERGE** until validated that Drupal changes A) successfully publish to the Outreach calendar and B) doesn't break anything else on FE
  - [x] Ticket FE changes needed to pick up the change
  - [x] Ticket merging the PR if it should not merge until after FE work is complete
- [x] Feature is behind a feature toggle
- [x] Appropriate test(s) added
- [x] Requires design review
- [x] Requires PM review
- [x] Requires a11y review

**Comment by wesrowe** [comment]
BE refinement:
- What do the role/permissions/section-assigments look like for this "outreach-only editor"?

**Comment by jilladams** [comment]
How to Events get associated with Listings: 
* Reverse entity lookup. 
* On the Event, currently, an Editor selects the appropriate list from "Where should the event be listed?" (`edit-field-listing`)

**Comment by jilladams** [comment]
Noting from planning: 
* Checkbox needs to map to the Listing selection, for the sake of FE, and we need to make that work with custom code or figure out for sure if/how it can
* Christia to ride along for learning

**Comment by thejordanwood** [comment]
@dsasser If the checkbox ends up being too tricky, I added an alternative design idea [in the Figma file](https://www.figma.com/file/8ZT1nnu8r2twbOlEhysnNp/Events?type=design&node-id=0%3A1&mode=design&t=RJSSNfasCgmeQDkL-1). This uses a multiselect modal similar to the benefit hub section in the [Video List content type](https://staging.cms.va.gov/node/add/media_list_videos).

**Comment by dsasser** [comment]
# Status Update 9/29/23


## Frontend

We unpacked how the FE is building event listings in content-build. The FE uses a graphql query, similar to the below abbreviated Graphql query to pull all events, grouped by event listing. The query shows how the data is being fetched by the FE, which is done by reverse entity referencing the `field_listing` field. Ultimately, we need to be adding additional values to this field, rather than adding a separate data source/field for the Outreach calendar, in order to avoid necessitating FE changes.

<details>

<summary>Graphql for building event listings</summary>

This graphql query is for a single 'event_list' (The National Outreach Calendar) for ease of testing:


```graphql
query MyQuery {
  nodeQuery(
        limit: 1000
        offset: 0
        sort: { field: "nid", direction:  ASC }
        filter: {
        conditions: [
          { field: "status", value: ["1"], enabled: true },
          { field: "type", value: ["event_listing"] },
          { field: "nid", value:"736"}
        ]
    
      })
  {
    count
        entities {
          ... on NodeEventListing {
            entityId
            entityLabel
            reverseFieldListingNode(limit: 50, filter: { conditions: [{ field: "status", value: "1", operator: EQUAL, enabled: true }, { field: "moderation_state", value: "archived", operator: NOT_EQUAL }, { field: "type", value: "event" }]}, sort: {field: "changed", direction: DESC}) {
              count
              entities {
                entityLabel
              }
            }
          }
        }
    }
  }
```

</details>

## Backend

We looked at two approaches for the backend. Both require first setting the `field_listing` field to support multiple values. By doing this, it immediately changes how the field is presented when editing, adding the 'multiple' property to the existing \<select\> element. This, in turn, changes the behavior of the select field from a dropdown to a multi-select element. This type of element does not have a great UX, since in order to select multiple items, a user would need to cmd+click/control+click each value.

![Screenshot 2023-09-29 at 12 41 29 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/8a6a4a33-b070-4678-bad2-d8de5ffac18f)

### Checkbox option

![Screenshot 2023-09-29 at 12 13 14 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/1e9764a3-1314-4a15-b3ea-a77657fa3d87)


The checkbox option is viable, providing we make a few changes:

- Add the new checkbox field (boolean) to the Event node
- Set the `field_listing` field storage settings to allow for multiple items (unlimited)
- In the form display settings, configure the `field_listing` field widget to show only a single element (screenshot below--this is provided by a contrib module, limited_field_widgets.
- Adding a form alter to control the visiblity of the new checkbox
- Adding a pre-save entity hook to add the Outreach calendar to the `field_listing`.
- A small amount of css to meet styles in Figma.

Pros: easy to implement
Cons: adds another field to the CMS; somewhat of a black box--Outreach Hub only users will have no awareness this is happening; the fact that `field_listing` contains more values than is present could lead to some confusion.

### Entity Browser option

![Screenshot 2023-09-29 at 12 21 10 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/7548e43a-7b60-4093-84fa-18873d55653e)

We also looked at using Entity Browser on the exiting `field_listing` field en lieu of an additional field. Since Entity Browser is used in many places in the site, there are no notable UX concerns. To accommodate automatically setting the National Outreach Calendar, we looked at using the 'default value' (see screenshot). While this does solve some of the checkbox concerns, it is very difficult to navigate setting the default value contextually based on user section.

Pros: easy to implement (in basic form); leveraged all over in the CMS (not a new paradigm); no 'hidden' event listing assignments.
Cons: very difficult to contextually apply the default value. After spending the better part of half-a day trying to work that out, I have made little progress.

**Comment by chri5tia** [comment]
Status Update 10/2/23
We sync'd up as a team and made a decision to go the checkbox route instead of the entity browser. Code is working and the next step is to write cypress tests and do functional testing.

**Comment by jilladams** [comment]
Thread re: change management: https://dsva.slack.com/archives/C52CL1PKQ/p1696357745318219

Need to talk test plan / roll out before this work merges.

**Comment by jilladams** [comment]
We need to: 
* create a test event with the new fix in tugboat
* Check and see what breadcrumbs do by default
* Vet that with Dave to determine if #13532 is blocking

**Comment by dsasser** [comment]
[Breadcrumb test ](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10089#issuecomment-1747420124)update:

❌  Adding an event to 2 event listings will cause, in the case of Lovell at least, a 404 to be returned in 1 of the 2 listing pages. I will be testing other facility types, as well as unpacking how the urls are being re-written, but this would suggest we have a launch blocking issue, ostensibly in vets-website.

## Tugboat
I rebuilt the tugboat we were using for demo'ing this issue, prior to doing any testing, to confirm that there were no errant content changes I wasn't expecting. During the rebuild, I noticed that the log was showing MANY broking link errors, and it was preventing the build from completing. Given this, I decided to do two things:

1) Setup a test locally, and
2) Setup a fresh tugboat, after merging in the latest code from main into the feature branch

The new tugboat instance is [here](https://vacms-10089-nat-outreach-cal-thu3raipnqtrntaczkhou6vtskp73ajc.demo.cms.va.gov/).

The latest content relase is still pending, so I moved onto test locally. I'll update here once I have a confirmed if the tugboat env works identically as locally (as expected).

## Local
I created a test event using the Lovell Health Care event listing, and the Outreach checkbox. The new event appeared on both listing pages (/outreach-and-events/events/ and /lovell-federal-health-care-va/events/ respectively).

From the Outreach and Events page, the link to the event was:  /lovell-federal-health-care/events/61945

Following this link gave:
<img width="1120" alt="Screenshot 2023-10-04 at 3 29 47 PM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/4400c61c-5325-4214-9137-9bc6a4de2bfe">

From the Lovell Events page, the link to the event was: /lovell-federal-health-care-va/events/61945

Following this link gave:
<img width="1203" alt="Screenshot 2023-10-04 at 3 30 20 PM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/6f499277-0fd4-47f0-bb7e-2942b69da658">

## Breadcrumb code
Breadcrumbs for Drupal pages are rendered with the `breadcrumbs.drupal.liquid` template, included from the `event.drupal.liquid` template.

<img width="995" alt="Screenshot 2023-10-04 at 3 32 56 PM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/a2f2dd8e-041e-4a15-9065-892f76de01fc">

In the breadcrumb template, this code is at play:
```
    {% if deriveBreadcrumbsFromUrl == true %}
      {% assign crumbs = entityUrl.breadcrumb | deriveLastBreadcrumbFromPath: title, entityUrl.path, replaceLastItem %}
    {% endif %}
``` 

This code uses the custom Liquid  filter `deriveLastBreadcrumbFromPath`, which leverages the url in order to create the breadcrumb.

**Comment by dsasser** [comment]
I tested the Alaska Heath Care facility, and the 404 mentioned above is _not_ present. The test event appears on on both the Alaska Health Care facility, and the Outreach and Events page. The breadcrumb, as expected, follows the URL:

Home -> Alaska health care Events -> Lorem Ipsum Test Event 

See:
https://web-thu3raipnqtrntaczkhou6vtskp73ajc.demo.cms.va.gov/alaska-health-care/events/
https://web-thu3raipnqtrntaczkhou6vtskp73ajc.demo.cms.va.gov/outreach-and-events/events/

So it would seem we have a Lovell-only issue with regards to posting a single Event to multiple Event Listings.

**Comment by jilladams** [comment]
#13532 is blocking merging this work. For now we are going to: add a feature toggle (just on/off) and hide the checkbox, in order to merge/ship. We will need a follow up ticket to either modify the toggle to show to a subset of users when it's on, or turn on, or just remove it.

**Comment by chri5tia** [comment]
# Status Update 10/5/23

As of before this morning's scrum update, cypress test was added and functional to test different scenarios regarding this feature.

Today, Daniel and I coworked to created a feature toggle for this outreach calendar checkbox due to the breadcrumb issue. As of right now there is something that needs to be debugged with it, so that is what I will be working on next between now and next scrum update.

Adding the feature toggle also means revising the cypress test, which means strategizing.

Currently cypress tests are added to existing step definition `i_create_a_node.js` which instructs how to fill out all the important form fields. The current events form code there saves the form. Below are ideas to improve this step definition to make it more versatile. In the meantime, we'll likely just need to add instructions to turn on the feature toggle before trying to create the node, but still thinking through the feasibility of it.

## Events Cypress Testing For the future
Consider changing the node create step definition to not include saving, removing the event part to `i_fill_in_node_details`, so that we can add different scenarios that all rely on filling out the required and other details without creating duplicitious test content; Add scenarios to events.feature, and similar:

```
  Scenario: Users can publish an event to the National Outreach Calendar
    Given I am logged in as a user with the "content_admin" role
    When I create an "event" node
    And I check the "Publish to the National Outreach Calendar" checkbox
    And I save the node
    Then the event node should be created successfully.
    And I should see "Outreach Calendar"

  Scenario: Users who can only publish to National Outreach Calendar do not see the "Publish to the National Outreach Calendar" checkbox
    Given I am logged in as a user with the "content_admin" role
    When my workbench access sections are set to "7"
    And I create an "event" node
    Then I should not see a "Publish to the National Outreach Calendar" checkbox
    When I save the node
    Then I should see "Outreach Calendar"
```

Also, then we can revise the first Scenario as such:

```
  Scenario: Log in and create an event.
    Given I am logged in as a user with the "content_admin" role
    When I create a "event" node
    And I save the node
    Then the event node should be created successfully.
```

We could add a step definition for `node_should_be_created_successfully` which checks for the success message and then goes back and checks for the revision log entry.

Also add `(a|an)` for create a node step.

**Comment by chri5tia** [comment]
- [ ] Also need to look at the `platform/file_upload.feature` cypress test which now fails due to changes to the events form.

**Comment by jilladams** [comment]
Poked at Lovell use case for sanity. Screensots in doobly doo. 

<details>

## VA Lovell listing

![Screenshot 2023-10-06 at 12 35 15 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/e5538fff-d1a1-4c65-93bd-4fab1c587df6)

## Tricare Lovell listing

![Screenshot 2023-10-06 at 12 37 27 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/205d68be-2743-4e06-94f5-307a5b602624)

## Outreach events

![Screenshot 2023-10-06 at 12 38 48 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/0d1dd548-7fe3-40fc-8bec-63a7cd389109)

## Event page, approach from VA Lovell

![Screenshot 2023-10-06 at 12 36 52 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/6d12859e-d157-4627-a2b4-7fef13fa69f1)

## Event page, approach from Tricare Lovell

![Screenshot 2023-10-06 at 12 37 57 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/7ca60f6c-96ca-4650-bf88-b516e682328f)

## Event page from Outreach events
Uses Federal health care VA breadcrumb. 

![Screenshot 2023-10-06 at 12 39 15 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/5902b133-58b0-4a05-8f4c-aa4ebc45565b)

</details>

(The inconsistencies in punctuation / capitalization between Event Listings and Events are the same in prod, so not part of our work here. Looking for a ticket and will file one if it doesn't exist.)

**Comment by chri5tia** [comment]
# Status Update 10/5/23

- Updated cypress test to include feature toggle under different user roles
- Fixed feature toggle so that the checkbox part of the form can be turned off when we don't want it
- [ ] This is ready to merge pending PO sign off, PM and a11y review

**Comment by chri5tia** [comment]
# Status update 10/9/23
- PR is merged

## Notes for future work
- Per @dsasser, consider changing step definition for feature toggle to run a drush command so that if it's used in the future, the feature file doesn't have to include steps to log in as admin and then log out and then log in as test user.

**Comment by dsasser** [comment]
@laflannery can you confirm (I think you did this in various places, but wanted to document here) that the CMS changes are good from an a11y perspective?

**Comment by laflannery** [comment]
@dsasser I I just reviewed on stage and have no issues. This is good on my end

**Comment by jilladams** [comment]
Noting:  I created https://github.com/department-of-veterans-affairs/va.gov-cms/issues/15627 to track any follow up required from talking to CAIA, and to track enabling the Feature toggle when the time comes.

**Comment by jilladams** [comment]
Work was merged to prod and worked fine on Events, but: broke other things. 
We are mid-trying to revert. 

#15641 tracks the related defect that will allow re-merge.
